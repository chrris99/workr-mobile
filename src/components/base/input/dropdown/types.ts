import { BaseControlledInputProps } from "@/components/base/input/types";
import { FieldValues } from "react-hook-form";

export type DropdownItem = {
  label: string;
  value: string;
  icon?: string;
};

export interface DropdownItemProps {
  item: DropdownItem;
  isSelected: boolean;
}

export interface DropdownInputProps<T extends FieldValues>
  extends BaseControlledInputProps<T> {
  data: DropdownItem[];
  label?: string;
  placeholder?: string;
  searchPlaceholder?: string;
}

export interface BottomSheetDropdownInputProps<T extends FieldValues>
  extends DropdownInputProps<T> {}
