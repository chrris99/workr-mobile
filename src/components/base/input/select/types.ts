import { FieldValue, FieldValues } from "react-hook-form";
import { IconName } from "../../../../design-system/icons/icons";
import { BaseControlledInputProps } from "../types";

export type SelectItem<T> = {
  value: T;
  iconName: IconName;
  displayText: string;
};

type BaseSelectItemProps<T> = {
  onChange: (...event: any[]) => void;
  item: SelectItem<T>;
};

export type SingleSelectItemProps<T> = BaseSelectItemProps<T> & {
  selectedItem: T;
  setSelectedItem: (value: T) => void;
};

export type MultiSelectItemProps<T> = BaseSelectItemProps<T> & {
  selectedItems: T[];
  setSelectedItems: (indexes: T[]) => void;
};

export type SelectItemProps<T> = {
  item: SelectItem<T>;
  isSelected: boolean;
  onPress: () => void;
};

export interface SelectInputProps<T extends FieldValues, V>
  extends BaseControlledInputProps<T> {
  options: SelectItem<V>[];
}
