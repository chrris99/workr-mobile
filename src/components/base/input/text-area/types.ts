import { BaseControlledInputProps } from "@/components/base/input/types";
import { FieldValues } from "react-hook-form";
import { TextInputProps } from "react-native";

export interface TextAreaProps<T extends FieldValues>
  extends BaseControlledInputProps<T>,
    Omit<TextInputProps, "multiline" | "numberOfLines"> {
  label?: string;
}

export interface BottomSheetTextAreaProps<T extends FieldValues>
  extends TextAreaProps<T> {}
