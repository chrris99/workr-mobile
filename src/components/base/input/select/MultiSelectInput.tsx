import { StyleSheet, View } from "react-native";
import { useState } from "react";
import { SelectInputProps } from "./types";
import { spacing } from "../../../../design-system/spacing/spacing";
import { Controller, FieldValues } from "react-hook-form";
import { MultiSelectItem } from "./MultiSelectItem";

export const MultiSelectInput = <T extends FieldValues, V>({
  control,
  name,
  rules,
  error,
  options,
}: SelectInputProps<T, V>) => {
  const [selectedItems, setSelectedItems] = useState<V[]>([]);

  const renderMultiSelectOptions = (onChange: () => void) =>
    options.map((item, index) => (
      <MultiSelectItem
        key={index}
        onChange={onChange}
        item={item}
        selectedItems={selectedItems}
        setSelectedItems={setSelectedItems}
      />
    ));

  return (
    <Controller
      control={control}
      rules={rules}
      name={name}
      render={({ field: { onChange, onBlur, value } }) => (
        <View style={styles.container}>
          {renderMultiSelectOptions(onChange)}
        </View>
      )}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    gap: spacing["spacing-3"],
  },
});
