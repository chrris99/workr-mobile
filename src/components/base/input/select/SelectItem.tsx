import { Checkbox } from "@/components/base/checkbox/Checkbox";
import { SelectItemProps } from "@/components/base/input/select/types";
import { colors } from "@/design-system/colors/colors";
import { Icon } from "@/design-system/icons/Icon";
import { spacing } from "@/design-system/spacing/spacing";
import Text from "@/design-system/typography/Text";
import { testId } from "@/utils/test/testId";
import { StyleSheet, View } from "react-native";
import { TouchableWithoutFeedback } from "react-native-gesture-handler";

export const SELECT_ITEM_TEST_IDS = {
  SELECT_ITEM: testId("select-item"),
} satisfies Record<string, string>;

export const SelectItem = <T,>({
  item,
  isSelected,
  onPress,
}: SelectItemProps<T>) => {
  const selectionStyle = isSelected ? styles.selected : styles.notSelected;

  return (
    <TouchableWithoutFeedback
      style={[styles.container, selectionStyle]}
      onPress={onPress}
      testID={SELECT_ITEM_TEST_IDS.SELECT_ITEM}
    >
      <View style={styles.contentContainer}>
        <Icon
          name={item.iconName}
          color={isSelected ? "primary-700" : "gray-900"}
        />
        <Text style={styles.textContainer}>{item.displayText}</Text>
        <Checkbox selected={isSelected} />
      </View>
    </TouchableWithoutFeedback>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingVertical: spacing["spacing-2"],
    paddingHorizontal: spacing["spacing-3"],
    borderRadius: spacing["spacing-2"],
  },
  selected: {
    backgroundColor: colors["primary-50"],
    borderColor: colors["primary-700"],
    borderWidth: 2,
  },
  notSelected: {
    borderColor: colors["gray-200"],
    borderWidth: 2,
  },
  contentContainer: {
    flexDirection: "row",
    gap: spacing["spacing-2"],
    alignItems: "center",
  },
  textContainer: {
    alignSelf: "center",
    flexGrow: 1,
  },
});
