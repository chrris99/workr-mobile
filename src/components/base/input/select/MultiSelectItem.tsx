import { View } from "react-native";
import { SelectItem } from "./SelectItem";
import { MultiSelectItemProps } from "./types";
import { testId } from "../../../../utils/testId";

export const MULTI_SELECT_ITEM_TEST_IDS = {
  ITEM: testId("multi-select-item"),
} satisfies Record<string, string>;

export const MultiSelectItem = <T,>({
  item,
  onChange,
  selectedItems,
  setSelectedItems,
}: MultiSelectItemProps<T>) => {
  const isSelected = selectedItems.includes(item.value);

  const onPress = () => {
    isSelected ? unselectItem() : selectItem();
  };

  const selectItem = () => {
    const newSelectedItems = [item.value, ...selectedItems];

    setSelectedItems(newSelectedItems);
    onChange(newSelectedItems);
  };

  const unselectItem = () => {
    const newSelectedItems = selectedItems.filter(
      (value) => value !== item.value
    );

    setSelectedItems(newSelectedItems);
    onChange(newSelectedItems);
  };

  return (
    <View testID={`${MULTI_SELECT_ITEM_TEST_IDS.ITEM}-${item.value}`}>
      <SelectItem item={item} isSelected={isSelected} onPress={onPress} />
    </View>
  );
};
