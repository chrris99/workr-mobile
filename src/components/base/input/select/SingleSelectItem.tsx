import { SingleSelectItemProps } from "./types";
import { SelectItem } from "./SelectItem";
import { View } from "react-native";
import { testId } from "../../../../utils/testId";

export const SINGLE_SELECT_ITEM_TEST_IDS = {
  ITEM: testId("single-select-item"),
} satisfies Record<string, string>;

export const SingleSelectItem = <T,>({
  item,
  onChange,
  selectedItem,
  setSelectedItem,
}: SingleSelectItemProps<T>) => {
  const isSelected = selectedItem === item.value;

  const onPress = () => {
    setSelectedItem(item.value);
    onChange(item.value);
  };

  return (
    <View testID={`${SINGLE_SELECT_ITEM_TEST_IDS.ITEM}-${item.value}`}>
      <SelectItem item={item} isSelected={isSelected} onPress={onPress} />
    </View>
  );
};
