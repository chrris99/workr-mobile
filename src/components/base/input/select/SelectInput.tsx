import { SingleSelectItem } from "@/components/base/input/select/SingleSelectItem";
import { SelectInputProps } from "@/components/base/input/select/types";
import { spacing } from "@/design-system/spacing/spacing";
import { useState } from "react";
import { Controller, FieldValues } from "react-hook-form";
import { StyleSheet, View } from "react-native";

export const SelectInput = <T extends FieldValues, V>({
  control,
  name,
  rules,
  error,
  options,
}: SelectInputProps<T, V>) => {
  const [selectedItem, setSelectedItem] = useState<V>(options[0].value);

  const renderOptions = (onChange: () => void) =>
    options.map((item, index) => (
      <SingleSelectItem
        key={index}
        onChange={onChange}
        item={item}
        selectedItem={selectedItem}
        setSelectedItem={setSelectedItem}
      />
    ));

  return (
    <Controller
      control={control}
      rules={rules}
      name={name}
      render={({ field: { onChange, onBlur, value } }) => (
        <View style={styles.container}>{renderOptions(onChange)}</View>
      )}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    gap: spacing["spacing-3"],
  },
});
