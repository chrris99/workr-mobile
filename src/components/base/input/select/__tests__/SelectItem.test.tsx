import {
  SELECT_ITEM_TEST_IDS,
  SelectItem,
} from "@/components/base/input/select/SelectItem";
import { Item } from "@/components/base/input/select/types";
import { fireEvent, render } from "@testing-library/react-native";

const ABS_ITEM: Item<string> = {
  value: "abs",
  iconName: "Abs",
  displayText: "Abs",
};

const mockOnPress = jest.fn();

describe("select item", () => {
  describe("in selected state", () => {
    it("should render correctly", () => {
      const component = render(
        <SelectItem item={ABS_ITEM} isSelected={true} onPress={mockOnPress} />
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe("in unselected state", () => {
    it("should render correctly", () => {
      const component = render(
        <SelectItem item={ABS_ITEM} isSelected={false} onPress={mockOnPress} />
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe("on press", () => {
    it("should call on press prop", () => {
      const { getByTestId } = render(
        <SelectItem item={ABS_ITEM} isSelected={false} onPress={mockOnPress} />
      );

      const selectItem = getByTestId(SELECT_ITEM_TEST_IDS.SELECT_ITEM);
      fireEvent.press(selectItem);

      expect(mockOnPress).toBeCalledTimes(1);
    });
  });
});
