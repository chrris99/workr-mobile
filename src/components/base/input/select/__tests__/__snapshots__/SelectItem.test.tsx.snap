// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select item in selected state should render correctly 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  delayLongPress={600}
  enabled={true}
  exclusive={true}
  handlerTag={1}
  handlerType="NativeViewGestureHandler"
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleColor={0}
  testID="workr:select-item"
  touchSoundDisabled={false}
>
  <View
    accessible={true}
    collapsable={false}
    style={
      {
        "backgroundColor": "#F9F5FF",
        "borderColor": "#6941C6",
        "borderRadius": 8,
        "borderWidth": 2,
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 8,
        }
      }
    >
      <Text
        style={
          [
            undefined,
            {
              "fontFamily": "Poppins_400Regular",
              "fontSize": 14,
              "lineHeight": 20,
            },
            {
              "color": "#101828",
            },
          ]
        }
      >
        Abs
      </Text>
      <Text
        style={
          [
            {
              "alignSelf": "center",
              "flexGrow": 1,
            },
            {
              "fontFamily": "Poppins_400Regular",
              "fontSize": 14,
              "lineHeight": 20,
            },
            {
              "color": "#101828",
            },
          ]
        }
      >
        Abs
      </Text>
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 8,
              "height": 16,
              "justifyContent": "center",
              "width": 16,
            },
            {
              "backgroundColor": "#6941C6",
              "borderColor": "#6941C6",
              "borderWidth": 1,
            },
          ]
        }
      >
        <Text
          style={
            [
              undefined,
              {
                "fontFamily": "Poppins_400Regular",
                "fontSize": 14,
                "lineHeight": 20,
              },
              {
                "color": "#101828",
              },
            ]
          }
        >
          Check
        </Text>
      </View>
    </View>
  </View>
</RNGestureHandlerButton>
`;

exports[`select item in unselected state should render correctly 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  delayLongPress={600}
  enabled={true}
  exclusive={true}
  handlerTag={2}
  handlerType="NativeViewGestureHandler"
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleColor={0}
  testID="workr:select-item"
  touchSoundDisabled={false}
>
  <View
    accessible={true}
    collapsable={false}
    style={
      {
        "borderColor": "#EAECF0",
        "borderRadius": 8,
        "borderWidth": 2,
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 8,
        }
      }
    >
      <Text
        style={
          [
            undefined,
            {
              "fontFamily": "Poppins_400Regular",
              "fontSize": 14,
              "lineHeight": 20,
            },
            {
              "color": "#101828",
            },
          ]
        }
      >
        Abs
      </Text>
      <Text
        style={
          [
            {
              "alignSelf": "center",
              "flexGrow": 1,
            },
            {
              "fontFamily": "Poppins_400Regular",
              "fontSize": 14,
              "lineHeight": 20,
            },
            {
              "color": "#101828",
            },
          ]
        }
      >
        Abs
      </Text>
      <View
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 8,
              "height": 16,
              "justifyContent": "center",
              "width": 16,
            },
            {
              "borderColor": "#D0D5DD",
              "borderRadius": 8,
              "borderWidth": 1,
            },
          ]
        }
      />
    </View>
  </View>
</RNGestureHandlerButton>
`;
